lib/src/position.dart

```diff
@@ -213,6 +213,208 @@ abstract class Position<T extends Position<T>> {
+      for (int rank = 0; rank < 8; rank++) {
+        final rankFilter = SquareSet.fromRank(rank).complement();
+        // If the square is behind or on this rank, the rank it will not contain the source pawn
+        if (turn == Side.white && rank >= squareRank(destination) ||
+            turn == Side.black && rank <= squareRank(destination)) {
+          filter = filter.intersect(rankFilter);
+        }
+      }
+      final source = (turn == Side.white) ? filter.last : filter.first;
+
+      // There are no valid candidates for the move
+      if (source == null) {
+        return null;
+      }
+
+      final move =
+          NormalMove(from: source, to: destination, promotion: promotionRole);
+      if (!isLegal(move)) {
+        return null;
+      }
+      return move;
+    }
+
+    // The final two moves define the destination
+    final destination = parseSquare(san.substring(san.length - 2));
+    if (destination == null) {
+      return null;
+    }
+
+    san = san.substring(0, san.length - 2);
+    if (isCapturing) {
+      // Invalid SAN
+      if (san[san.length - 1] != "x") {
+        return null;
+      }
+      san = san.substring(0, san.length - 1);
+    }
+
+    // For non-pawn moves, the first character describes a role
+    final role = Role.fromChar(san[0]);
+    if (role == null) {
+      return null;
+    }
+    if (role == Role.pawn) {
+      return null;
+    }
+    san = san.substring(1);
+
+    final colorFilter = board.bySide(turn);
+    final roleFilter = board.byRole(role);
+    SquareSet filter = colorFilter.intersect(roleFilter);
+
+    // The remaining characters disambiguate the moves
+    if (san.length > 2) {
+      return null;
+    }
+    if (san.length == 2) {
+      final sourceSquare = parseSquare(san);
+      if (sourceSquare == null) {
+        return null;
+      }
+      final squareFilter = SquareSet.fromSquare(sourceSquare);
+      filter = filter.intersect(squareFilter);
+    }
+    if (san.length == 1) {
+      final sourceCharacter = san.codeUnits[0];
+      if (oneIndex <= sourceCharacter && sourceCharacter <= eightIndex) {
+        final rank = sourceCharacter - oneIndex;
+        final rankFilter = SquareSet.fromRank(rank);
+        filter = filter.intersect(rankFilter);
+      } else if (aIndex <= sourceCharacter && sourceCharacter <= hIndex) {
+        final file = sourceCharacter - aIndex;
+        final fileFilter = SquareSet.fromFile(file);
+        filter = filter.intersect(fileFilter);
+      } else {
+        return null;
+      }
+    }
+
+    Move? move;
+    for (final square in filter.squares) {
+      final candidateMove = NormalMove(from: square, to: destination);
+      if (!isLegal(candidateMove)) {
+        continue;
+      }
+      if (move == null) {
+        move = candidateMove;
+      } else {
+        // Ambiguous notation
+        return null;
+      }
+    }
+
+    if (move == null) {
+      return null;
+    }
+
+    return move;
```

Comment:
I don't think it supports `DropMove` and Crazyhouse variant right now. It should support them.